sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,123 +32,206 @@
+  /**
+   * Interface for calculating and tracking pane info state.
+   */
+  public interface PaneInfoTrackerIface {
+    void clear(StateAccessor<?> state);
+
+    ReadableState<PaneInfo> getNextPaneInfo(
+        ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal);
+
+    void storeCurrentPaneInfo(ReduceFn<?, ?, ?, ?>.Context context, PaneInfo currentPane);
+
+    boolean hasState();
   }
 
   @VisibleForTesting
   static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
       StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
 
-  public void clear(StateAccessor<?> state) {
-    state.access(PANE_INFO_TAG).clear();
+
+  public static PaneInfoTrackerIface globalPaneInfoTracker() {
+    return new GlobalPaneInfoTracker();
+  }
+
+  public static PaneInfoTrackerIface localPaneInfoTracker(TimerInternals timerInternals) {
+    return new LocalPaneInfoTracker(timerInternals);
   }
 
   /**
-   * Return a ({@link ReadableState} for) the pane info appropriate for {@code context}. The pane
-   * info includes the timing for the pane, who's calculation is quite subtle.
-   *
-   * @param isFinal should be {@code true} only if the triggering machinery can guarantee
-   * no further firings for the
+   * Track pane info for Global window.
    */
-  public ReadableState<PaneInfo> getNextPaneInfo(
-      ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal) {
-    final Object key = context.key();
-    final ReadableState<PaneInfo> previousPaneFuture =
-        context.state().access(PaneInfoTracker.PANE_INFO_TAG);
-    final Instant windowMaxTimestamp = context.window().maxTimestamp();
-
-    return new ReadableState<PaneInfo>() {
-      @Override
-      public ReadableState<PaneInfo> readLater() {
-        previousPaneFuture.readLater();
-        return this;
-      }
+  private static class GlobalPaneInfoTracker implements PaneInfoTrackerIface {
+    @Override
+    public void clear(StateAccessor<?> state) {
+      // Nothing to clear.
+    }
 
-      @Override
-      public PaneInfo read() {
-        PaneInfo previousPane = previousPaneFuture.read();
-        return describePane(key, windowMaxTimestamp, previousPane, isFinal);
-      }
-    };
-  }
+    @Override
+    public ReadableState<PaneInfo> getNextPaneInfo(
+        ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal) {
+      return new ReadableState<PaneInfo>() {
+        @Override
+        public ReadableState<PaneInfo> readLater() {
+          // Nothing to read.
+          return this;
+        }
+
+        @Override
+        public PaneInfo read() {
+          // Since this is for the Global window:
+          //  - never first.
+          //  - always early.
+          //  - random nonce for speculative index.
+          //  - always zero for non-speculative index.
+          return PaneInfo.createPane(false, isFinal, isFinal  ? Timing.ON_TIME : Timing.EARLY,
+              ThreadLocalRandom.current().nextLong(), 0L);
```

Comment:
Yes, but before I go polishing I wanted confidence we can get everyone to agree.

I'll send  email now.